File: subsystems/outtake/ShooterSubsystem.java
package org.firstinspires.ftc.teamcode.subsystems.outtake;

import org.firstinspires.ftc.teamcode.utils.SMO.SMOFilter;

import dev.nextftc.control.ControlSystem;
import dev.nextftc.control.KineticState;
import dev.nextftc.control.builder.FilterBuilder;
import dev.nextftc.control.feedback.FeedbackType;
import dev.nextftc.control.feedback.PIDCoefficients;
import dev.nextftc.core.commands.Command;
import dev.nextftc.core.commands.utility.InstantCommand;
import dev.nextftc.core.subsystems.Subsystem;
import dev.nextftc.hardware.impl.MotorEx;

public class ShooterSubsystem implements Subsystem {
    public static final ShooterSubsystem INSTANCE = new ShooterSubsystem();

    private final MotorEx shootMotor = new MotorEx("Shooter");

    private double targetVelocity = 1.0;

    public static double Ls = 0.0, Lv = 0.0, La = 0.0, kSqu = 0.0, kI = 0.0, kD = 0.0;
    private ControlSystem controlSystem;

    private ShooterSubsystem() {
        SMOFilter velSMO = new SMOFilter(FeedbackType.VELOCITY, Ls, Lv, La);
        controlSystem = ControlSystem.builder()
                .velFilter(filter -> filter.custom(velSMO))
                .velSquID(new PIDCoefficients(kSqu, kI, kD))
                .build();
    }

    @Override
    public void periodic() {
        shootMotor.setPower(
                controlSystem.calculate(
                    new KineticState(0, targetVelocity)  // todo: if not work, consider pos != 0
                )
        );
    }

    public Command setVelocity(double targetVelocity) {
        return new InstantCommand(() -> this.targetVelocity = targetVelocity);
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/outtake/ShooterSubsystem.kt
package org.firstinspires.ftc.teamcode.subsystems.outtake

import com.bylazar.configurables.annotations.Configurable
import com.bylazar.telemetry.PanelsTelemetry.telemetry
import com.qualcomm.robotcore.util.ElapsedTime
import dev.nextftc.control.ControlSystem
import dev.nextftc.control.KineticState
import dev.nextftc.control.builder.controlSystem
import dev.nextftc.control.feedback.FeedbackType
import dev.nextftc.control.feedback.PIDCoefficients
import dev.nextftc.control.feedforward.BasicFeedforwardParameters
import dev.nextftc.core.commands.Command
import dev.nextftc.core.commands.groups.ParallelGroup
import dev.nextftc.core.subsystems.Subsystem
import dev.nextftc.core.units.Angle
import dev.nextftc.hardware.controllable.RunToState
import dev.nextftc.hardware.controllable.RunToVelocity
import dev.nextftc.hardware.impl.MotorEx
import dev.nextftc.hardware.powerable.SetPower
import org.firstinspires.ftc.teamcode.subsystems.outtake.turret.TurretPhiSubsystem
import org.firstinspires.ftc.teamcode.subsystems.outtake.turret.TurretThetaSubsystem.targetTheta
import org.firstinspires.ftc.teamcode.utils.SMO.BasicSMOParameters
import org.firstinspires.ftc.teamcode.utils.SMO.SMOFilter
import java.util.function.Supplier


@Configurable
object ShooterSubsystem : Subsystem {
//    private val motor1 = MotorEx("shooter1");
    private val motor2 = MotorEx("shooter2").reversed().brakeMode();

    @JvmField var ffCoefficients = BasicFeedforwardParameters(0.0, 0.0, 1.0);
    @JvmField var pidCoefficients = PIDCoefficients(0.0, 0.01, 0.0)
    @JvmField var shootingCoefficients = PIDCoefficients(0.0, 0.01, 0.0)

    public val velocity: Double
        get() {
            return controller.lastMeasurement.velocity;
        }

    private val controller: ControlSystem;
    private val shootingController: ControlSystem;

    init {
        controller = controlSystem {
            basicFF(ffCoefficients)
            velPid(pidCoefficients)
        }

        // todo: tune shooting controller
        shootingController = controlSystem {
            basicFF(ffCoefficients)
            velPid(shootingCoefficients)
        }
    }

    fun setMotorPowers(power: Double) {
//        motor1.power = power;
        motor2.power = power;
    }

    fun setControllerGoals(velocity: Double) {
        controller.goal = KineticState(velocity=velocity);
        shootingController.goal = KineticState(velocity=velocity);
    }

    class On(private val velocity: Double): Command() {
        override val isDone = true;

        override fun start() {
            setControllerGoals(velocity)
        }
    }

    var off = On(0.0);

    class Manual(
        private val shooterPower: Supplier<Double>
    ) : Command() {
        override val isDone = false;

        override fun update() {
            setMotorPowers(shooterPower.get())
        }
    }

    class AutoAim(
        private val dxy: Double,
        private val powerByDistance: (Double) -> Double,  // get by running curve of best fit on collected data
    ) : Command() {
        override val isDone = true;

        init {
            requires(ShooterSubsystem)
        }

        override fun start() {
            setControllerGoals(powerByDistance(dxy))
        }
    }

    var isShooting = false;
    var lastPos = 0.0;
    var elapsedTime: ElapsedTime = ElapsedTime();
    override fun periodic() {
        val power = (if (isShooting) shootingController else controller).calculate(
            motor2.state.times(-1.0)
        ).coerceIn(0.0, 1.0);

        setMotorPowers(power);

        val measuredVel = (motor2.currentPosition - lastPos)/elapsedTime.time();
        lastPos = motor2.currentPosition;
        elapsedTime.reset()

        telemetry.addData("power", power)

        telemetry.addData("vel measured", measuredVel)
        telemetry.addData("vel est", controller.lastMeasurement.velocity)
        telemetry.addData("vel ref", controller.reference.velocity)
        telemetry.addData("vel goal", controller.goal.velocity)

//        telemetry.addData("pos measured 1", motor1.currentPosition)
        telemetry.addData("pos measured 2", motor2.currentPosition)
        telemetry.addData("pos est", -controller.lastMeasurement.position)
        telemetry.addData("pos ref", controller.reference.position)
    }
}